### PR TITLE
Add Support for Python Upgrade Rules

### PR DESCRIPTION
**Pull request by thsparks**
Need this so we can update python code when we make a breaking change. I mostly just copied whatever we were doing for typescript and tried to do it for python as well.

### CODE REVIEW

File: pxtcompiler/simpledriver.ts [hunk 1]
```diff
@@ -194,16 +197,24 @@ namespace pxt {
...
         for (let fn of Object.keys(opts.fileSystem)) {
-            if (fn.indexOf("/") == -1 && U.endsWith(fn, ".ts")) {
-                const ts = opts.fileSystem[fn]
-                const ts2 = pxt.patching.patchJavaScript(version, ts)
-                if (ts != ts2) {
-                    pxt.debug(`applying TS patch to ${fn}`)
-                    opts.fileSystem[fn] = ts2
+            if (fn.indexOf("/") == -1 && U.endsWith(fn, extension)) {
+                const initial = opts.fileSystem[fn]
+                const patched = pxt.patching.patchPython(version, initial)
```
[the remark this answers is not in the document]
Comment (srietkerk) [reply pxtcompiler/simpledriver.ts:214]: Or should there just also be `patchJavascript` here?

File: webapp/src/compiler.ts [hunk 2]
```diff
@@ -953,6 +958,37 @@ function upgradeFromTSAsync(): Promise<UpgradeResult> {
...
+
+    const patchedFiles: pxt.Map<string> = {};
+    pxt.Util.values(project.files).filter(isPyFile).forEach(file => {
+        const patched = pxt.patching.patchPython(targetVersion, file.content);
+        if (patched != file.content) {
+            patchedFiles[file.name] = patched;
+        }
+    });
+
+    pxt.debug("Applying upgrades to Python")
```
Comment (srietkerk) [inline webapp/src/compiler.ts:974]: Do we want to leave this in?

File: pxtcompiler/simpledriver.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwunderl) [reply pxtcompiler/simpledriver.ts:214]: yeah, looks like this needs something like a `const patchFunction = extension == ".ts" ? pxt.patching.patchJavaScript : pxt.patching.patchPython;` then call that here to me?

[the remark this answers is not in the document]
Comment (thsparks) [reply pxtcompiler/simpledriver.ts:222]: I opted to just keep them separate. Seemed simpler at this point...

File: webapp/src/compiler.ts [hunk 3]
```diff
@@ -953,6 +958,37 @@ function upgradeFromTSAsync(): Promise<UpgradeResult> {
         });
 }
 
+function upgradeFromPythonAsync(): Promise<UpgradeResult> {
```
Comment (riknoll) [inline webapp/src/compiler.ts:961]: async/await-ify this please

**Review by riknoll — APPROVED** [review]

**Review by jwunderl — APPROVED** [review]

**Review by srietkerk — APPROVED** [review]
Nice!

**Pull request merged**